File: backend/app/routers/study.py
import json
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from .. import models, schemas
from ..database import get_db
from ..dependencies import get_current_user

router = APIRouter(prefix="/api", tags=["study"])

units_path = Path(__file__).resolve().parent.parent / "data" / "units.json"
units_data: list[dict] = json.loads(units_path.read_text(encoding="utf-8"))
units_by_id: dict[str, dict] = {u["id"]: u for u in units_data}

details_path = Path(__file__).resolve().parent.parent / "data" / "unit_details.json"
details_by_id: dict[str, dict] = json.loads(details_path.read_text(encoding="utf-8"))

details_extra_path = Path(__file__).resolve().parent.parent / "data" / "unit_details_extra.json"
details_extra_by_id: dict[str, dict] = (
    json.loads(details_extra_path.read_text(encoding="utf-8")) if details_extra_path.exists() else {}
)

details_new_path = Path(__file__).resolve().parent.parent / "data" / "unit_details_new.json"
details_new_by_id: dict[str, dict] = (
    json.loads(details_new_path.read_text(encoding="utf-8")) if details_new_path.exists() else {}
)

curriculum_path = Path(__file__).resolve().parent.parent / "data" / "curriculum.json"
curriculum_raw: dict = (
    json.loads(curriculum_path.read_text(encoding="utf-8")) if curriculum_path.exists() else {"units": []}
)


@router.get("/units", response_model=list[schemas.UnitSummary])
def list_units(current_user: models.User = Depends(get_current_user)):
    if curriculum_raw.get("units"):
        lesson_ids: list[str] = []
        seen: set[str] = set()
        for raw_unit in list(curriculum_raw.get("units") or []):
            for lesson_id in list(raw_unit.get("lessons") or []):
                lesson_key = str(lesson_id)
                if lesson_key in units_by_id and lesson_key not in seen:
                    lesson_ids.append(lesson_key)
                    seen.add(lesson_key)
        return [{"id": units_by_id[i]["id"], "title": units_by_id[i]["title"]} for i in lesson_ids]

    return [{"id": u["id"], "title": u["title"]} for u in units_data]


@router.get("/curriculum", response_model=schemas.Curriculum)
def get_curriculum(current_user: models.User = Depends(get_current_user)):
    _ = current_user
    out_units: list[dict] = []
    for raw_unit in list(curriculum_raw.get("units") or []):
        unit_id = raw_unit.get("id")
        title = raw_unit.get("title")
        if not isinstance(unit_id, str) or not unit_id:
            continue
        if not isinstance(title, str) or not title:
            continue

        lessons: list[dict] = []
        for lesson_id in list(raw_unit.get("lessons") or []):
            lesson = units_by_id.get(str(lesson_id))
            if not lesson:
                continue
            lessons.append({"id": lesson["id"], "title": lesson["title"]})

        out_units.append(
            {
                "id": unit_id,
                "title": title,
                "description": raw_unit.get("description"),
                "lessons": lessons,
            }
        )

    return {"units": out_units}


@router.get("/units/{unit_id}", response_model=schemas.UnitDetail)
def get_unit(unit_id: str, current_user: models.User = Depends(get_current_user)):
    unit = units_by_id.get(unit_id)
    if not unit:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Unit not found")

    learn = dict(unit.get("learn") or {})
    base = details_by_id.get(unit_id) or {}
    extra = details_extra_by_id.get(unit_id) or {}
    new = details_new_by_id.get(unit_id) or {}

    learn["details"] = [
        *list(base.get("details") or []),
        *list(extra.get("details") or []),
        *list(new.get("details") or []),
    ]
    learn["summary"] = [
        *list(base.get("summary") or []),
        *list(extra.get("summary") or []),
        *list(new.get("summary") or []),
    ]

    return {
        "id": unit["id"],
        "title": unit["title"],
        "learn": learn,
        "quiz": {"question": unit["quiz"]["question"], "choices": unit["quiz"]["choices"]},
    }


@router.post("/units/{unit_id}/answer", response_model=schemas.UnitAnswerResponse)
def submit_answer(
    unit_id: str,
    payload: schemas.UnitAnswerRequest,
    current_user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    unit = units_by_id.get(unit_id)
    if not unit:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Unit not found")

    correct_answer = int(unit["quiz"]["answer"])
    correct = payload.answer == correct_answer

    progress = (
        db.query(models.UnitProgress)
        .filter(models.UnitProgress.user_id == current_user.id, models.UnitProgress.unit_id == unit_id)
        .first()
    )
    if progress is None:
        progress = models.UnitProgress(user_id=current_user.id, unit_id=unit_id)
        db.add(progress)

    progress.attempts = int(progress.attempts or 0) + 1
    if correct:
        progress.correct_answers = int(progress.correct_answers or 0) + 1
        progress.completed = True

    db.commit()
    db.refresh(progress)

    return schemas.UnitAnswerResponse(correct=correct, progress=schemas.UnitProgressOut.model_validate(progress))


@router.get("/progress", response_model=list[schemas.UnitProgressOut])
def get_progress(
    current_user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    items = db.query(models.UnitProgress).filter(models.UnitProgress.user_id == current_user.id).all()
    return [schemas.UnitProgressOut.model_validate(p) for p in items]

File: frontend-only/src/api/study.ts
import { client } from './client';
import { OFFLINE_MODE } from '../config';
import { getCurriculum as getLocalCurriculum } from '../data/curriculum';
import { getLessonById, getLessons } from '../data/lessons';
import { getUnitById, getUnits } from '../data/units';
import { getCurrentLanguage } from '../i18n';
import { getOfflineSessionEmail, getOfflineProgress, updateOfflineProgress } from '../offline';

export type UnitSummary = { id: string; title: string };

export type CurriculumLesson = {
  id: string;
  title: string;
};

export type CurriculumUnit = {
  id: string;
  title: string;
  description?: string | null;
  lessons: CurriculumLesson[];
};

export type CurriculumSection = {
  id: string;
  title: string;
  units: CurriculumUnit[];
};

export type Curriculum = { sections: CurriculumSection[] };

export type UnitDetail = {
  id: string;
  title: string;
  lessonId: string;
  quiz: { question: string; choices: string[] };
};

export type LessonSummary = { id: string; title: string };

export type LessonDetail = {
  id: string;
  title: string;
  learn: { text: string; code?: string | null; details?: string[]; summary?: string[] };
  quizIds: string[];
};

export type UnitProgress = {
  unit_id: string;
  attempts: number;
  correct_answers: number;
  completed: boolean;
};

export type UnitAnswerResponse = {
  correct: boolean;
  progress: UnitProgress;
};

type SubmitOptions = { offline?: boolean };

const getLang = () => getCurrentLanguage();

const requireOfflineEmail = async () => {
  const email = await getOfflineSessionEmail();
  if (!email) {
    throw new Error('Not authenticated');
  }
  return email;
};

const listUnitsOffline = (): UnitSummary[] => {
  const lang = getLang();
  const localizedUnits = getUnits(lang);
  const unitById = new Map(localizedUnits.map((unit) => [unit.id, unit]));
  const curriculum = getLocalCurriculum(lang);
  const sections = Array.isArray((curriculum as any).sections) ? (curriculum as any).sections : [];
  const unitsFromSections = sections.flatMap((section: any) =>
    Array.isArray(section?.units) ? section.units : [],
  );
  const unitsFromRoot = Array.isArray((curriculum as any).units) ? (curriculum as any).units : [];
  const curriculumLessonIds = [...unitsFromSections, ...unitsFromRoot].flatMap((unit: any) => {
    if (!Array.isArray(unit?.lessons)) return [];
    return unit.lessons.map((lesson: any) => (typeof lesson === 'string' ? lesson : lesson?.id));
  });
  if (!curriculumLessonIds.length) {
    return localizedUnits.map((unit) => ({ id: unit.id, title: unit.title }));
  }
  const out: UnitSummary[] = [];
  const seen = new Set<string>();
  for (const lessonId of curriculumLessonIds) {
    const unit = unitById.get(lessonId);
    if (!unit || seen.has(unit.id)) continue;
    seen.add(unit.id);
    out.push({ id: unit.id, title: unit.title });
  }
  return out.length ? out : localizedUnits.map((unit) => ({ id: unit.id, title: unit.title }));
};

const listLessonsOffline = (): LessonSummary[] => {
  const lang = getLang();
  return getLessons(lang).map((lesson) => ({ id: lesson.id, title: lesson.title }));
};

const getUnitOffline = (unitId: string): UnitDetail => {
  const lang = getLang();
  const unit = getUnitById(lang, unitId);
  if (!unit) {
    throw new Error('Unit not found');
  }
  return {
    id: unit.id,
    title: unit.title,
    lessonId: unit.lessonId,
    quiz: { question: unit.quiz.question, choices: unit.quiz.choices },
  };
};

const getLessonOffline = (lessonId: string): LessonDetail => {
  const lang = getLang();
  const lesson = getLessonById(lang, lessonId);
  if (!lesson) {
    throw new Error('Lesson not found');
  }
  return {
    id: lesson.id,
    title: lesson.title,
    learn: lesson.learn,
    quizIds: lesson.quizIds,
  };
};

const getCurriculumOffline = (): Curriculum => {
  const lang = getLang();
  const curriculum = getLocalCurriculum(lang);
  const localizedUnits = getUnits(lang);
  const unitById = new Map(localizedUnits.map((unit) => [unit.id, unit]));
  return {
    sections: curriculum.sections.map((section) => ({
      id: section.id,
      title: section.title,
      units: section.units.map((unit) => ({
        id: unit.id,
        title: unit.title,
        description: unit.description ?? null,
        lessons: unit.lessons
          .map((lessonId) => {
            const lesson = unitById.get(lessonId);
            return lesson ? { id: lesson.id, title: lesson.title } : null;
          })
          .filter((lesson): lesson is UnitSummary => Boolean(lesson)),
      })),
    })),
  };
};

const getProgressOffline = async (): Promise<UnitProgress[]> => {
  const email = await requireOfflineEmail();
  return getOfflineProgress(email);
};

const submitAnswerOffline = async (unitId: string, answer: number): Promise<UnitAnswerResponse> => {
  const unit = getUnitById(getLang(), unitId);
  if (!unit) {
    throw new Error('Unit not found');
  }
  const correct = answer === unit.quiz.answer;
  const email = await requireOfflineEmail();
  const progress = await updateOfflineProgress(email, unitId, correct);
  return { correct, progress };
};

export const listUnits = async () => {
  return listUnitsOffline();
};

export const getCurriculum = async () => {
  return getCurriculumOffline();
};

export const getUnit = async (unitId: string) => {
  return getUnitOffline(unitId);
};

export const listLessons = async () => {
  return listLessonsOffline();
};

export const getLesson = async (lessonId: string) => {
  return getLessonOffline(lessonId);
};

export const getProgress = async () => {
  if (OFFLINE_MODE) return getProgressOffline();
  return client.get<UnitProgress[]>('/api/progress').then((r) => r.data);
};

export const submitAnswer = async (unitId: string, answer: number, options?: SubmitOptions) => {
  if (OFFLINE_MODE || options?.offline) return submitAnswerOffline(unitId, answer);
  return client
    .post<UnitAnswerResponse>(`/api/units/${encodeURIComponent(unitId)}/answer`, { answer })
    .then((r) => r.data);
};
